### PR TITLE
Expose internal_dynamic_variant_config to Evaluations

### PR DESCRIPTION
**Pull request by anndvision**
This PR enables evaluations to test dynamically-generated variant candidates without adding them to the config file first. This is essential for optimization workflows (like GEPA) that need to evaluate variant candidates on-the-fly.

### Core Infrastructure (Rust)
- Added optional `dynamic_variant_config: Option<UninitializedVariantInfo>` field to `EvaluationCoreArgs`
- Threaded the parameter through the evaluation pipeline (`run_evaluation_core_streaming`, `infer_datapoint`)

### Python Client
- Exposed `dynamic_variant_config` parameter in both sync and async `experimental_run_evaluation` methods
- Added proper deserialization from Python dict to `UninitializedVariantInfo`
- Fixed lifetime annotations for async FFI bindings

### Node.js Client
- Updated to include `dynamic_variant_config: None` in evaluation construction

### Testing
- Added 3 Rust tests in `evaluations/tests/tests.rs`:
  - `test_evaluation_with_no_dynamic_config` - baseline test with `None`
  - `test_evaluation_with_dynamic_variant` - test with dynamic variant provided
  - `test_dynamic_variant_overrides_config_lookup` - test that dynamic config takes precedence
- Added 2 Python tests in `clients/python/tests/test_evaluation.py`:
  - `test_sync_run_evaluation_with_dynamic_variant`
  - `test_async_run_evaluation_with_dynamic_variant`

All tests pass successfully.

## TODO

- [x] Make the added tests more meaningful and comprehensive (follow-up after PR submission)
  - Current tests verify basic functionality but should be enhanced to test edge cases
  - Should verify that dynamic variants actually override config lookups in observable ways
  - Should test error handling for invalid dynamic variant configurations

----

> [!IMPORTANT]
> Add support for dynamic variant configurations in evaluations, enabling on-the-fly testing without config file changes, with updates to Rust core, Python, and Node.js clients, and comprehensive testing.
> 
>   - **Behavior**:
>     - Add `dynamic_variant_config` to `EvaluationCoreArgs` in Rust, allowing dynamic variant testing without config file changes.
>     - Modify `run_evaluation_core_streaming` and `infer_datapoint` to handle `dynamic_variant_config`.
>     - Ensure dynamic variants override static config lookups.
>   - **Python Client**:
>     - Add `dynamic_variant_config` parameter to `experimental_run_evaluation` methods in `TensorZeroGateway` and `AsyncTensorZeroGateway`.
>     - Implement deserialization from Python dict to `UninitializedVariantInfo`.
>     - Fix lifetime annotations for async FFI bindings.
>   - **Node.js Client**:
>     - Update to include `dynamic_variant_config: None` in evaluation construction.
>   - **Testing**:
>     - Add Rust tests in `tests.rs` for dynamic variant handling and precedence over static config.
>     - Add Python tests in `test_evaluation.py` for sync and async evaluations with dynamic variants.
>     - Ensure tests cover error handling for invalid configurations.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=tensorzero%2Ftensorzero&utm_source=github&utm_medium=referral)<sup> for f1a558c74ee17b6dd18e60f8e4f3d4bd76d849a8. You can [customize](https://app.ellipsis.dev/tensorzero/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: clients/python/src/lib.rs [hunk 1]
```diff
@@ -1102,14 +1105,36 @@ impl TensorZeroGateway {
...
+                    return Err(PyValueError::new_err(
+                        "Either 'variant_name' or 'dynamic_variant_config' must be provided.",
+                    ));
+                }
+                (Some(_), Some(_)) => {
+                    return Err(PyValueError::new_err(
+                        "Cannot specify both 'variant_name' and 'dynamic_variant_config'. \
+                        When using a dynamic variant, provide only 'dynamic_variant_config'.",
+                    ));
+                }
```
[the remark this answers is not in the document]
Comment (virajmehta) [reply clients/python/src/lib.rs]: the helper should take both args, deserialize, and return a Result<EvaluationCoreVariant> or whatevr that type is called.

File: evaluations/tests/tests.rs [hunk 2]
```diff
@@ -2211,3 +2218,245 @@ async fn test_query_skips_staled_datapoints() {
...
+        result.is_ok(),
+        "Evaluation with dynamic variant should succeed"
+    );
+
+    let result = result.unwrap();
+    assert!(result.run_info.num_datapoints > 0);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_dynamic_variant_overrides_config_lookup() {
```
[the remark this answers is not in the document]
Comment (virajmehta) [reply evaluations/tests/tests.rs]: or remove if the types make this impossible

**Review by virajmehta — CHANGES_REQUESTED** [review]

**Review by virajmehta — CHANGES_REQUESTED** [review]
tiny nit and I'll stamp
